src/util/animate.ts

```diff
@@ -89,15 +182,15 @@ export function animate(options = {}) {
   };
   runningAnimations.push(context);
 
-  const runner = function (timestamp) {
+  const runner = function (timestamp: number) {
     const start = timestamp || +new Date(),
       finish = start + duration,
       isMany = Array.isArray(startValue),
```

Comment:
sometimes i see this done:

function isValueMany(value: number | number[]): value is number[] {
  return Array.isArray(value);
}

We could do `isMany = isValueMany(startValue) && isValueMany(endValue)`

And this could avoid having to then do the as number[] in the ternary.
It would clarify we verified those are arrays

Reply:
Hm I tried this and it seems that it only confuses the typing system more because it doesn't clarify that if they aren't arrays they are numbers, at least according to typescript. Or maybe I did it wrong lol.

Reply:
I wanted to suggest that too but couldn't get it to work as well.
Type assertion don't negate unfortunately so you will need to write 2 type assertion functions.

Reply:
I got it wrong before
now is good

```tsx
// https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates

function isMany<T, S extends T[]>(t:T|S): t is S {
    return Array.isArray(t)
}


function go(b:number|number[]) {
    if(isMany(b)){
        b.push(1)
    } else {
        b+1;
    }
}
```

[Playground Link](https://www.typescriptlang.org/play?ts=4.8.2#code/PTAEAsBdIBwZwFwgO6oHSQJ4wKZwMYBOAljJADYCGAdgOZoD2htwAJg-nMODawEYMGAa2AAmYNUqFCDZMTpooAW3IBiAK5x5tALRZcOmIRyti+SpDwBYAFC2AZuur5IxBtVDE4AWRqYAPAAqADSgAMqgOAAeltSscKCBANoAugB8ABSQCIEAPmEAlAigkJ4JEQDetqA1oMaQ6oQeAILSlJhoXq2E7VkFtgC+tsM2js6u7qC0DBl8CNTqSnw4hLkLSyupBaBVNrWe9hlevtSYswUFu-v7fGgwmuAZAIz9e7UDkeRwODvV1zV8ADUTwA3H8akMbJCgA)

Reply:
well then typescript sucks.
It can't done both are arrays, and it can't do is array or number depending on this boolean.
Ts sucks

Reply:
Even better


```tsx
// https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates

function isMany<T, S extends T extends unknown[]?T:T[]>(t:T|S): t is S {
    return Array.isArray(t)
}


function go(b:number|number[]) {
    if(isMany(b)){
        b.push(1)
    } else {
        b+1;
    }
}

```

[Playground Link](https://www.typescriptlang.org/play?ts=4.8.2#code/PTAEAsBdIBwZwFwgO6oHSQJ4wKZwMYBOAljJADYCGAdgOZoD2htwAJg-nMODawEYMGAa2AAmYNUqFCDZMTpooAW3IBiAK5x5tALRZcOmIRyti+SpDwBYAFC2AZuur5IxBtVDE4AWRqYAPAAqADSgAMqgOAAeltSscKCBkTE4cQlOQtSy1ADaALoA-IEIgfkAfAAUkCUAPmEAlAigkJ4JEQDetqDdoMaQ6oQeAILSlJhoXiOEY1X1tgC+tks2js6u7qC0DBV8CNTqSnw4hDX7h8f59aCdNj2e9hVevtSYO-X1N3d3fGgwmuAVACMc1uPXmkXIcBw1y6X26fAA1ICANyw7qLGwYoA)
```

[Playground Link](https://www.typescriptlang.org/play?ts=4.8.2#code/PTAEAsBdIBwZwFwgO6oHSQJ4wKZwMYBOAljJADYCGAdgOZoD2htwAJg-nMODawEYMGAa2AAmYNUqFCDZMTpooAW3IBiAK5x5tALRZcOmIRyti+SpDwBYAFC2AZuur5IxBtVDE4AWRqYAPAAqADSgAMqgOAAeltSscKCBkTE4cQlOQtSy1ADaALoA-IEIgfkAfAAUkCUAPmEAlAigkJ4JEQDetqDdoMaQ6oQeAILSlJhoXiOEY1X1tgC+tks2js6u7qC0DBV8CNTqSnw4hDX7h8f59aCdNj2e9hVevtSYO-X1N3d3fGgwmuAVACMc1uPXmkXIcBw1y6X26fAA1ICANyw7qLGwYoA)

Reply:
it's not the first time we comment in a race condition... 😅

Reply:
@Lazauya I would like implemented

Reply:
I don't mind doing it.
And make all the interfaces generic `AnimationOptions<T exnteds number | number[]>`